ext/standard/proc_open.c

```diff
@@ -646,199 +885,56 @@ PHP_FUNCTION(proc_open)
-				} else {
-					descriptors[ndesc].parentend = newpipe[0];
-					descriptors[ndesc].childend = newpipe[1];
-				}
-#ifdef PHP_WIN32
-				/* don't let the child inherit the parent side of the pipe */
-				descriptors[ndesc].parentend = dup_handle(descriptors[ndesc].parentend, FALSE, TRUE);
-#endif
-				descriptors[ndesc].mode_flags = descriptors[ndesc].mode & DESC_PARENT_MODE_WRITE ? O_WRONLY : O_RDONLY;
-#ifdef PHP_WIN32
-				if (Z_STRLEN_P(zmode) >= 2 && Z_STRVAL_P(zmode)[1] == 'b')
-					descriptors[ndesc].mode_flags |= O_BINARY;
-#endif
-
-			} else if (strcmp(Z_STRVAL_P(ztype), "file") == 0) {
-				zval *zfile, *zmode;
-				php_socket_t fd;
-				php_stream *stream;
-
-				descriptors[ndesc].mode = DESC_FILE;
-
-				if ((zfile = zend_hash_index_find(Z_ARRVAL_P(descitem), 1)) != NULL) {
-					if (!try_convert_to_string(zfile)) {
-						goto exit_fail;
-					}
-				} else {
-					zend_value_error("Missing file name parameter for 'file'");
+			} else if (strcmp(ZSTR_VAL(ztype), "file") == 0) {
+				if (get_string_parameter(&zfile, descitem, 1, "file name parameter for 'file'") == FAILURE) {
 					goto exit_fail;
 				}
-
-				if ((zmode = zend_hash_index_find(Z_ARRVAL_P(descitem), 2)) != NULL) {
-					if (!try_convert_to_string(zmode)) {
-						goto exit_fail;
-					}
-				} else {
-					zend_value_error("Missing mode parameter for 'file'");
+				if (get_string_parameter(&zmode, descitem, 2, "mode parameter for 'file'") == FAILURE) {
 					goto exit_fail;
 				}
 
-				/* try a wrapper */
-				stream = php_stream_open_wrapper(Z_STRVAL_P(zfile), Z_STRVAL_P(zmode),
-						REPORT_ERRORS|STREAM_WILL_CAST, NULL);
-
-				/* force into an fd */
-				if (stream == NULL || FAILURE == php_stream_cast(stream,
-							PHP_STREAM_CAST_RELEASE|PHP_STREAM_AS_FD,
-							(void **)&fd, REPORT_ERRORS)) {
+				if (set_proc_descriptor_to_file(&descriptors[ndesc], zfile, zmode) == FAILURE) {
 					goto exit_fail;
 				}
-
-#ifdef PHP_WIN32
-				descriptors[ndesc].childend = dup_fd_as_handle((int)fd);
-				_close((int)fd);
-
-				/* simulate the append mode by fseeking to the end of the file
-				this introduces a potential race-condition, but it is the best we can do, though */
-				if (strchr(Z_STRVAL_P(zmode), 'a')) {
-					SetFilePointer(descriptors[ndesc].childend, 0, NULL, FILE_END);
-				}
-#else
-				descriptors[ndesc].childend = fd;
-#endif
-			} else if (strcmp(Z_STRVAL_P(ztype), "redirect") == 0) {
+			} else if (strcmp(ZSTR_VAL(ztype), "redirect") == 0) {
 				zval *ztarget = zend_hash_index_find_deref(Z_ARRVAL_P(descitem), 1);
-				struct php_proc_open_descriptor_item *target = NULL;
-				php_file_descriptor_t childend;
-
 				if (!ztarget) {
 					zend_value_error("Missing redirection target");
 					goto exit_fail;
 				}
 				if (Z_TYPE_P(ztarget) != IS_LONG) {
-					zend_value_error("Redirection target must be an integer");
+					zend_value_error("Redirection target must be an integer; got %s", zend_get_type_by_const(Z_TYPE_P(ztarget)));
```

Comment:
```suggestion
					zend_value_error("Redirection target must be of type int, %s given", zend_get_type_by_const(Z_TYPE_P(ztarget)));
```

is the usual phrasing.

Reply:
Done.

Reply:
I think this part is not done yet.

Reply:
I think I was still running `make test` and hadn't pushed yet when you made this comment.